marketplace-builder/src/builder.rs

```diff
@@ -0,0 +1,390 @@
+        instance_state: NodeState,
+        validated_state: ValidatedState,
+        hotshot_events_api_url: Url,
+        hotshot_builder_apis_url: Url,
+        max_api_timeout_duration: Duration,
+        buffered_view_num_count: usize,
+        maximize_txns_count_timeout_duration: Duration,
+        base_fee: FeeAmount,
+        bid_config: Option<BidConfig>,
+        solver_api_url: Url,
+    ) -> anyhow::Result<Self> {
+        tracing::info!(
+            address = %builder_key_pair.fee_account(),
+            ?bootstrapped_view,
+            %tx_channel_capacity,
+            %event_channel_capacity,
+            ?max_api_timeout_duration,
+            buffered_view_num_count,
+            ?maximize_txns_count_timeout_duration,
+            "initializing builder",
+        );
+
+        let (mut senders, receivers) =
+            marketplace_builder_core::service::broadcast_channels(event_channel_capacity.get());
+
+        senders.transactions.set_capacity(tx_channel_capacity.get());
+
+        // builder api request channel
+        let (req_sender, req_receiver) =
+            broadcast::<MessageType<SeqTypes>>(event_channel_capacity.get());
+
+        let (genesis_payload, genesis_ns_table) =
+            Payload::from_transactions([], &validated_state, &instance_state)
+                .await
+                .expect("genesis payload construction failed");
+
+        let builder_commitment = genesis_payload.builder_commitment(&genesis_ns_table);
+
+        let vid_commitment = {
+            let payload_bytes = genesis_payload.encode();
+            vid_commitment(&payload_bytes, GENESIS_VID_NUM_STORAGE_NODES)
+        };
+
+        // create the global state
+        let global_state: GlobalState<SeqTypes> = GlobalState::<SeqTypes>::new(
+            req_sender,
+            senders.transactions.clone(),
+            vid_commitment,
+            bootstrapped_view,
+            bootstrapped_view,
+            buffered_view_num_count as u64,
+        );
+
+        let global_state = Arc::new(RwLock::new(global_state));
+        let global_state_clone = global_state.clone();
+
+        let builder_state = BuilderState::<SeqTypes>::new(
+            BuiltFromProposedBlock {
+                view_number: bootstrapped_view,
+                vid_commitment,
+                leaf_commit: fake_commitment(),
+                builder_commitment,
+            },
+            &receivers,
+            req_receiver,
+            Vec::new() /* tx_queue */,
+            global_state_clone,
+            node_count,
+            maximize_txns_count_timeout_duration,
+            base_fee
+                .as_u64()
+                .context("the base fee exceeds the maximum amount that a builder can pay (defined by u64::MAX)")?,
+            Arc::new(instance_state),
+            Duration::from_secs(60),
+            Arc::new(validated_state),
+        );
+
+        // spawn the builder event loop
+        async_spawn(async move {
```

Comment:
Not sure if ignoring all the handles from these async tasks is okay.

Reply:
Due to how the builder-core works, this handle will actually eventually fork and exit, so we're fine with it exiting.

Reply:
How about adding this as a comment?